In [3]:
import pandas as pd
market_df = pd.read_csv("market_fact.csv")
print(market_df)

        Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08   
...        ...      ...       ...        ...        ...       ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
0                 23   -30.51           3.60                 0.56  
1                 13     4.56      

In [5]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [5]:
print(market_df[2:7])

     Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   
5  Ord_5446   Prod_6  SHP_7608  Cust_1818   164.02      0.03              23   
6    Ord_31  Prod_12    SHP_41    Cust_26    14.76      0.01               5   

    Profit  Shipping_Cost  Product_Base_Margin  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38  
5   -47.64           6.15                 0.37  
6     1.32           0.50                 0.36  


In [7]:
market_df.Sales

0        136.8100
1         42.2700
2       4701.6900
3       2337.8900
4       4233.1500
          ...    
8394    2841.4395
8395     127.1600
8396     243.0500
8397    3872.8700
8398     603.6900
Name: Sales, Length: 8399, dtype: float64

In [11]:
market_df[['Sales', 'Profit']]

Sales   Profit
0      136.8100   -30.51
1       42.2700     4.56
2     4701.6900  1148.90
3     2337.8900   729.34
4     4233.1500  1219.87
...         ...      ...
8394  2841.4395   374.63
8395   127.1600   -74.03
8396   243.0500   -70.85
8397  3872.8700   565.34
8398   603.6900   131.39

[8399 rows x 2 columns]

#### Positon and Label Based indexing in Pandas

In [14]:
# Position based indexing
help(pd.DataFrame.iloc)

Help on property:

    Purely integer-location based indexing for selection by position.

    .. deprecated:: 2.2.0

       Returning a tuple from a callable is deprecated.

    ``.iloc[]`` is primarily integer position based (from ``0`` to
    ``length-1`` of the axis), but may also be used with a boolean
    array.

    Allowed inputs are:

    - An integer, e.g. ``5``.
    - A list or array of integers, e.g. ``[4, 3, 0]``.
    - A slice object with ints, e.g. ``1:7``.
    - A boolean array.
    - A ``callable`` function with one argument (the calling Series or
      DataFrame) and that returns valid output for indexing (one of the above).
      This is useful in method chains, when you don't have a reference to the
      calling object, but would like to base your selection on
      some value.
    - A tuple of row and column indexes. The tuple elements consist of one of the
      above inputs, e.g. ``(0, 1)``.

    ``.iloc`` will raise ``IndexError`` if a requested indexer is
    o

In [16]:
market_df.iloc[2,4]

4701.69

In [30]:
market_df.iloc[1:5]

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [32]:
#label based indexing
help(pd.DataFrame.loc)

Help on property:

    Access a group of rows and columns by label(s) or a boolean array.

    ``.loc[]`` is primarily label based, but may also be used with a
    boolean array.

    Allowed inputs are:

    - A single label, e.g. ``5`` or ``'a'``, (note that ``5`` is
      interpreted as a *label* of the index, and **never** as an
      integer position along the index).
    - A list or array of labels, e.g. ``['a', 'b', 'c']``.
    - A slice object with labels, e.g. ``'a':'f'``.

      .. warning:: Note that contrary to usual python slices, **both** the
          start and the stop are included

    - A boolean array of the same length as the axis being sliced,
      e.g. ``[True, False, True]``.
    - An alignable boolean Series. The index of the key will be aligned before
      masking.
    - An alignable Index. The Index of the returned selection will be the input.
    - A ``callable`` function with one argument (the calling Series or
      DataFrame) and that returns valid outpu

In [34]:
market_df.loc[5]

Ord_id                  Ord_5446
Prod_id                   Prod_6
Ship_id                 SHP_7608
Cust_id                Cust_1818
Sales                     164.02
Discount                    0.03
Order_Quantity                23
Profit                    -47.64
Shipping_Cost               6.15
Product_Base_Margin         0.37
Name: 5, dtype: object

In [52]:
market_df.set_index('Ord_id', inplace=True)

KeyError: "None of ['Ord_id'] are in the columns"

In [58]:
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

In [66]:
market_df.loc['Ord_5446', ['Sales', 'Profit']]

Sales   Profit
Ord_id                    
Ord_5446   136.81   -30.51
Ord_5446  4701.69  1148.90
Ord_5446   164.02   -47.64

In [72]:
market_df.loc[['Ord_5446','Ord_5406', 'Ord_5485'], 'Sales':'Profit']

Sales  Discount  Order_Quantity   Profit
Ord_id                                              
Ord_5446   136.81      0.01              23   -30.51
Ord_5446  4701.69      0.00              26  1148.90
Ord_5446   164.02      0.03              23   -47.64
Ord_5406    42.27      0.01              13     4.56
Ord_5485  4233.15      0.08              35  1219.87

In [86]:
market_df.loc[:,[True, False, True, True, True, True, False, False, True]]

Prod_id    Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                              
Ord_5446  Prod_16  Cust_1818   136.8100      0.01              23   
Ord_5406  Prod_13  Cust_1818    42.2700      0.01              13   
Ord_5446   Prod_4  Cust_1818  4701.6900      0.00              26   
Ord_5456   Prod_6  Cust_1818  2337.8900      0.09              43   
Ord_5485  Prod_17  Cust_1818  4233.1500      0.08              35   
...           ...        ...        ...       ...             ...   
Ord_5353   Prod_4  Cust_1798  2841.4395      0.08              28   
Ord_5411   Prod_6  Cust_1798   127.1600      0.10              20   
Ord_5388   Prod_6  Cust_1798   243.0500      0.02              39   
Ord_5348  Prod_15  Cust_1798  3872.8700      0.03              23   
Ord_5459   Prod_6  Cust_1798   603.6900      0.00              47   

          Product_Base_Margin  
Ord_id                         
Ord_5446                 0.56  
Ord_5406                 0.54  
Ord_5446                 0.59  
Ord_5456                 0.37  
Ord_5485                 0.38  
...                       ...  
Ord_5353                 0.59  
Ord_5411                 0.37  
Ord_5388                 0.40  
Ord_5348                 0.62  
Ord_5459                 0.38  

[8399 rows x 6 columns]

In [88]:
market_df.Sales>3000

Ord_id
Ord_5446    False
Ord_5406    False
Ord_5446     True
Ord_5456    False
Ord_5485     True
            ...  
Ord_5353    False
Ord_5411    False
Ord_5388    False
Ord_5348     True
Ord_5459    False
Name: Sales, Length: 8399, dtype: bool

In [90]:
market_df.loc[market_df['Sales']>3000,:]

Prod_id   Ship_id    Cust_id       Sales  Discount  Order_Quantity  \
Ord_id                                                                         
Ord_5446   Prod_4  SHP_7610  Cust_1818   4701.6900      0.00              26   
Ord_5485  Prod_17  SHP_7664  Cust_1818   4233.1500      0.08              35   
Ord_4725   Prod_4  SHP_6593  Cust_1641   3410.1575      0.10              48   
Ord_4743   Prod_2  SHP_6615  Cust_1641   4072.0100      0.01              43   
Ord_2207  Prod_11  SHP_3093   Cust_839   3364.2480      0.10              15   
...           ...       ...        ...         ...       ...             ...   
Ord_3593   Prod_3  SHP_4974  Cust_1274  12073.0600      0.03              39   
Ord_3593  Prod_15  SHP_4975  Cust_1274   6685.0500      0.09              25   
Ord_2624   Prod_4  SHP_3591  Cust_1006   4924.1350      0.07              28   
Ord_2722   Prod_1  SHP_3731  Cust_1006   3508.3300      0.04              21   
Ord_5348  Prod_15  SHP_7469  Cust_1798   3872.8700      0.03              23   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446  1148.90           2.50                 0.59  
Ord_5485  1219.87          26.30                 0.38  
Ord_4725  1137.91           0.99                 0.55  
Ord_4743  1675.98           0.99                 0.56  
Ord_2207  -693.23          61.76                 0.78  
...           ...            ...                  ...  
Ord_3593  5081.87          19.99                 0.38  
Ord_3593  1653.60          24.49                  NaN  
Ord_2624  1049.54           8.99                 0.58  
Ord_2722  -546.98          35.00                 0.85  
Ord_5348   565.34          30.00                 0.62  

[1359 rows x 9 columns]

In [96]:
market_df.loc[(market_df.Sales>2000) | (market_df.Sales < 4000) & (market_df.Profit>100), ['Sales','Discount', 'Profit']]

Sales  Discount   Profit
Ord_id                                
Ord_5446  4701.6900      0.00  1148.90
Ord_5456  2337.8900      0.09   729.34
Ord_5485  4233.1500      0.08  1219.87
Ord_4725  3410.1575      0.10  1137.91
Ord_4743  4072.0100      0.01  1675.98
...             ...       ...      ...
Ord_2722  3508.3300      0.04  -546.98
Ord_1833   611.1600      0.04   100.22
Ord_5353  2841.4395      0.08   374.63
Ord_5348  3872.8700      0.03   565.34
Ord_5459   603.6900      0.00   131.39

[3009 rows x 3 columns]

#### Practice Problems

1. Given the dataset -> https://query.data.world/s/vBDCsoHCytUSLKkLvq851k2b8JOCkF do the following:
   i. Using iloc index the dataframe to print all the rows of the columns at index 3,4,5.
   ii. Using loc function print out all the columns and rows from 2 to 20 of the 'df' dataset.
   iii. Print all the columns and the rows where 'area' is greater than 0, 'wind' is greater than 1 and the 'temp' is greater than 15.
   iv. Using the same conditions as above, sort the resulting DataFrame by temp in descending order and display the top 5 rows.
   v. Using the df dataset, filter the rows where the RH (Relative Humidity) is less than 30 and rain is equal to 0. Then, sort the DataFrame by wind in ascending order.

2. Create a pandas dataframe using the following data: 

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
    'Age': [25, 30, 35, 40, 45],
    'City': ['New York', 'Los Angeles', 'Chicago', 'Houston', 'Phoenix']
}

Answer the following:
a. Write the code to display all the column names and data type of each column
b. Print the first and last two rows using position-based indexing.
c. Print the data for 'David' and 'Eve' using label-based slicing. 




3. Given a dataframe:
ipl18 = pd.DataFrame({'Team': ['SRH', 'CSK', 'KKR', 'RR', 'MI', 'RCB', 'KXIP', 'DD'],
                         'Matches': [14, 14, 14, 14, 14, 14, 14, 14],
                         'Won': [9, 9, 8, 7, 6, 6, 6, 5],
                         'Lost': [5, 5, 6, 7, 8, 8, 8, 9],
                         'Tied': [0, 0, 0, 0, 0, 0, 0, 0],
                         'N/R': [0, 0, 0, 0, 0, 0, 0, 0],
                         'Points': [18, 18, 16, 14, 12, 12, 12, 10],
                         'NRR': [0.284, 0.253, -0.070, -0.250, 0.317, 0.129, -0.502, -0.222],
                         'For': [2230, 2488, 2363, 2130, 2380, 2322, 2210, 2297],
                         'Against': [2193, 2433, 2425, 2141, 2282, 2383, 2259, 2304]},
                         index = range(1,9)
          
   Answer the following:
   a. Part A: Using label-based indexing (loc), write a Python program to:

        i. Select and print the rows for teams that have a NRR greater than 0.
        ii. Select and print the rows for teams that have won exactly 6 matches but have a Points score less than 14.
   
   b. Part B: Using position-based indexing (iloc), select and print:

        i. The Team, Won, and Points columns for the last 5 teams in the DataFrame.
        ii. All rows except the first two, displaying only the Matches and For columns.
   
   c. Part C: Sorting and Custom Indexing

       i. Sort the teams based on the number of Points in descending order. Display the sorted DataFrame.
       ii.Sort the DataFrame by the number of Won matches in descending order, and further sort by Points in descending order. 
             )